src/main/scala/tech/cryptonomic/conseil/Harpoon.scala

```diff
@@ -0,0 +1,153 @@
+import tech.cryptonomic.conseil.util.{DatabaseUtil, PureLogging}
+import tech.cryptonomic.conseil.config.HarpoonAppConfig
+import scala.concurrent.ExecutionContext
+import tech.cryptonomic.conseil.util.IOUtil
+import slick.jdbc.{ResultSetConcurrency, ResultSetType}
+import slick.jdbc.PostgresProfile.api._
+
+/** Runs an independent process that scans existing operations
+  * recorded from the chainm to look for reveal or activation
+  * and thus mark the corresponding Account record
+  */
+object Harpoon extends IOApp with PureLogging {
+
+  /** A type alias to provide state management on top of IO.
+    * The state threaded through corresponds to the actual
+    * process input value (AccountsTaggingProcess.ProcessorInput i.e. Option[BlockLevel]),
+    * which allows to start processing from that point onward
+    */
+  type IOState[T] = StateT[IO, Option[BlockLevel], T]
+
+  /* Reference to the database to use, we need write access to update accounts.
+   * A reference is guaranteed to be released after use, avoiding leaks
+   */
+  private val dbAccess: Resource[IO, Database] =
+    Resource.fromAutoCloseable(IO(DatabaseUtil.lorreDb))
+
+  // private lazy val dbHandle: Database = DatabaseUtil.lorreDb
+
+  /* This is the main entrypoint, the return type is wrapped in IO and provides an ExitCode for the process
+   * Described in cats-effect apis: https://typelevel.org/cats-effect/api/cats/effect/IOApp.html
+   */
+  override def run(args: List[String]): IO[ExitCode] = {
+
+    //we read custom conf, usually provided from the environment
+    val conf = HarpoonAppConfig
+
+    //we create the processor stage, with the custom types needed
+    val tagging = AccountsTaggingProcess[IOState]
+
+    //we use the database resource, which will be acquired before and released after
+    dbAccess.use { implicit db =>
+      implicit val ctx = scala.concurrent.ExecutionContext.global
+
+      /* Implement any dependency needed to run the processing. */
+      implicit val serviceApis = injectServices(tagging)
+
+      /* now we can call the processing stage */
+      val processCycle = (processedTopLevel: tagging.ProcessorInput) =>
+        tagging.process
+          .run(processedTopLevel)
+          .flatTap {
+            case (level, flagged) =>
+              logger.pureLog[IO](
+                _.info(
+                  "This Harpoon processing cycle marked {} accounts, up to level {}.",
+                  flagged.fold("an unknown number of")(String.valueOf),
+                  level.fold("unspecified")(String.valueOf)
+                )
+              )
+          }
+          .handleErrorWith(
+            err =>
+              logger.pureLog[IO](_.error("Harpoon processing failed to complete", err)) *>
+                  IO.pure((processedTopLevel, 0)) //we reset to the previous state as to retry next round
+          )
+
+      //a wait period before running the process once more
+      val waitRound =
+        IO.shift *>
```

Comment:
Now that I think though, is there a reason to shift if you're only doing this once at a time? I don't see how switching contexts would help.

Reply:
I'm using a single execution context for the whole application, which uses fs2 streams.
If the stream schedule tasks on different threads, I can feel assured that the main loop will do his job by yielding control to other threads, if needed (consider that we're doing IO on those too, such as logging).

Reply:
I'm just not sure there's a point if everything is sequential :P